SFTP/SSH as remote Backup location

maretodoric opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Taking advantage of 2025.1 of Backup locations being extendable by design, I've developed this integration that will allow user to store the backup to remote host via SSH/SFTP.

Once backup is triggered and completed, backup will be automatically uploaded to configurable host to a path specified by user.

This integration uses `paramiko` as I've noticed `paramiko` is already installed (at least on my system).
However, if required - i can rewrite to use AsyncSSH.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36973

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: Created PR: https://github.com/home-assistant/home-assistant.io/pull/36973

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

maretodoric committed:
Update requirements file via 'script.gen_requirements_all'

maretodoric committed:
first commit

maretodoric committed:
Merge branch 'dev' into dev

maretodoric committed:
Use defined variables instead of re-using `user_input.get`.
Adapt calling of `SSHClient.get_identifier` method.

maretodoric committed:
Include `backup_location` attribute in `SFTPConfigEntryData`

maretodoric committed:
Add `SFTPBackupAgent.unique_id` attribute to match the pattern of `SSHClient.get_identifier`.

maretodoric committed:
Accept `backup_location` argument for `SSHClient`.
Remove the need for `remote_path` argument in `SSHClient.get_identifier` method.

maretodoric committed:
Merge branch 'dev' into dev

zweckj reviewed (changes requested):
Since backups are a critical component and backup files can be quite large even on low powered systems, I'd highly prefer AsyncSSH to the byte chunking. 

This is also missing a [`quality_scale.yaml`](https://developers.home-assistant.io/docs/core/integration-quality-scale/checklist) file and reach at least bronze.

Lastly, be prepared you will also need to add 100% test coverage for both config flow and backup.py (but you can also do that once we agreed on the rest of the implementation)

maretodoric commented:
> Since backups are a critical component and backup files can be quite large even on low powered systems, I'd highly prefer AsyncSSH to the byte chunking.

Thanks for reviewing.
I have re-written it to use AsyncSSH, I'm testing it locally before i push changes.

> This is also missing a [`quality_scale.yaml`](https://developers.home-assistant.io/docs/core/integration-quality-scale/checklist) file and reach at least bronze.

I'll work on quality scale, this is first integration that I'm contributing to, quality scale is quite new to me and will take me some time.

> Lastly, be prepared you will also need to add 100% test coverage for both config flow and backup.py (but you can also do that once we agreed on the rest of the implementation)

It should not be a problem, i will work on it but i do need some clarification on test requirement.

For best possible test results, i could try to spawn a temporary SSH Server using `asyncssh.create_server` function. I would test for available ports in range between 1024 - 65535 and bind to first available port on localhost. However, will test environment allow for such action?

Additionally, if tests are executed as root user, that could pose a security risk, regardless of it listening on localhost. At that point, testing will become a lot more complicated :)
Any recommendations on how to proceed with tests in such cases or if you do not prefer me to run temporary SSH Server?

zweckj commented:
> For best possible test results, i could try to spawn a temporary SSH Server using `asyncssh.create_server` function. I would test for available ports in range between 1024 - 65535 and bind to first available port on localhost. However, will test environment allow for such action?
> 
> Additionally, if tests are executed as root user, that could pose a security risk, regardless of it listening on localhost. At that point, testing will become a lot more complicated :) Any recommendations on how to proceed with tests in such cases or if you do not prefer me to run temporary SSH Server?

That shouldn't be necessary as the lib will probably already abstract that. What we usually do in tests is mock (more or less) the entire library away (but definitely anything that'd do IO) so we are only testing the interaction of HA with the library. If Check the tests for onedrive or google_drive, that'll make it easier to understand.

maretodoric committed:
Merge branch 'dev' into dev

maretodoric committed:
Update integration to use asyncssh instead of paramiko.

maretodoric committed:
Update manifest to include `after_dependencies`, `integration_type`, `quality_scale` and update requirements to use `asyncssh` instead of paramiko.

maretodoric committed:
Update strings.

maretodoric committed:
Add quality scale.

maretodoric committed:
Add test cases.

maretodoric committed:
Remove unused SSHServer class - will not mock SSH Server.

maretodoric committed:
Perform ruff formatting and fix docstrings.

maretodoric committed:
Update codeowners and requirements with `script.gen_requirements_all` and `script.hassfest`.

maretodoric committed:
Merge branch 'dev' into dev

maretodoric commented:
It's done.
- Code re-written to use asyncssh,
- Test cases added, config flow is at 100% coverage and backup.py is at 99%.
- [quality_scale.yaml](https://developers.home-assistant.io/docs/core/integration-quality-scale/checklist) added - currently at bronze.

Additionally, I've added icon image to brands repo.

maretodoric committed:
Re-format with ruff

maretodoric committed:
Hide password in `SFTPConfigEntryData` string representation.
Trigger reauth flow if setup fails due to known exceptions.

maretodoric committed:
Change initialization log line to `DEBUG` leve instead of `INFO`.

maretodoric committed:
Implement reauth flow.

maretodoric committed:
Move `USER_INPUT` dict to conftest.

maretodoric committed:
Add test case for reauth flow.

maretodoric committed:
Add test case for reauth flow during setup stage.

maretodoric committed:
Add strings for `reauth_confirm` step and `abort` stage.

maretodoric committed:
Fix module docstring.

maretodoric committed:
Add test cases for `backup_sftp.util` module.

maretodoric committed:
Change log level to debug.

maretodoric committed:
Fix docstring information - we're not using API.

maretodoric committed:
Bump quality scale to silver.

maretodoric committed:
Move `create_tar_bytes` to conftest.

maretodoric committed:
Add client test case.

maretodoric commented:
I've pushed some new changes.
When debug logging is active, user password will not be printed into the log (for security purposes) instead, it will be masked.
Bumped quality scale to silver and improved test cases coverage.

maretodoric committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/backup_sftp/util.py:
```diff
@@ -0,0 +1,101 @@
+    return {
+        "backup_id": cfg["slug"],
+        "database_included": database_included,
+        "date": cfg["date"],
+        "extra_metadata": cfg["extra"],
+        "homeassistant_included": homeassistant_included,
+        "homeassistant_version": homeassistant_version,
+        "name": cfg["name"],
+        "protected": cfg["protected"],
+        "addons": [
+            AddonInfo(name=addon["name"], slug=addon["slug"], version=addon["version"])
+            for addon in cfg["addons"]
+        ],
+        "folders": [Folder(f) for f in cfg["folders"]],
+    }
+
+
+def process_tar_from_adapter(adapter: AsyncSSHFileWrapper, file: str) -> dict:
+    """Open ./backup.json from backup file archive.
+
+    KeyError is raised if file does not exist in archive (not a valid backup)
+    and same exception is raised if any values from config files are missing
+    ...again, indicating that the mentioned .tar file may not be home assistant backup.
+    """
+    with tarfile.open(mode="r", fileobj=adapter) as t:
+        # Force the IDE to treat 't' as a TarFile:
+        tar = cast(tarfile.TarFile, t)
+
+        try:
+            fileobj = tar.getmember("./backup.json")
+        except KeyError:
+            return None
+
+        try:
+            cfg = json.load(tar.extractfile(fileobj))
+        except Exception as e:
+            raise AssertionError(
+                f"Provided file ({file}) cannot be loaded as json."
+            ) from e
+    return get_backup_metadata(cfg)
```
so the metadata handling is not something we can do. We cannot extract the tar files. You'll probably need to just work with `.metadata.json` files like synology DSM does it. 

The other thing is, you shouldn't touch the `AgentBackup` it's considered a blob. You can do things like
```python
{
  "metadata_version": 1,
  "somekey": "bla",
  "agent_backup": json.dumps(AgentBackup.to_dict())
}
```
but don't work on it's attributes

maretodoric replied:
Honestly, it would be a lot easier to store metadata in some file and read from it rather than reading it from tar file.
I've originally wanted to store it in a separate file but i was concerned with:

1. Unnecessarily doubling the inode counts if i store metadata file per backup archive (in case someone stores ton of backups, not sure why would anyone do it but still, it should be taken into account)
2. Possible race condition when reading/writing metadata file if i store all files metadata in a single file.

Fix for first problem would be to apply second solution.
Fix for second solution would be to implement, like vim has, a swap file (.metadata.json.swp) and do not read/write if that file exists. Or to possibly read/write metadata to database. sqlite in default home assistant installation or other database if someone (like myself) configured another database.

What would be your thoughts ? I'm more than happy to discuss.

zweckj replied:
So the easy one first: With "touch" I mean: Don't try to construct anything. Dump it and read it. Don't extract keys or assume you have an understanding what's inside.

I'm not so much worried about inodes, that is a limitation we can document. Don't use a single "database" file, that's too brittle and if it break we'll break all backups.

maretodoric replied:
Done, every backup now has a single metadata file to read the info from, rather than reading tar object.

zweckj commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,349 @@
+                    )
+                    continue
+                metadata["size"] = (await rfile.stat()).size
+                backup = AgentBackup(**metadata)
```
it's also a bit weird that we are mixing specific/unspecific types in this class. Most of this code probably better moves to the backup

maretodoric replied:
Original idea was to de-clutter `backup.py`, to allow it to be more readable and to more easily perform unit tests (in terms to possibly gain better coverage in `backup.py`). Btw, what do you mean by mixing specific/unspecific types? I can add type hint however considering it's only assigned in one place in this function, with specific class (`AgentBackup`), kind of didn't see a point in adding type hint.

zweckj replied:
What I mean with specific: If this is a helper class it shouldn't throw backup specifc errors and if it does it's a good indication this better be in backup.py.

zweckj replied:
In general we should keep helper classes as short as really needed also to avoid falsify code coverage

maretodoric replied:
It's done, hopefully. Feel free to review and get back to me.

maretodoric committed:
Update homeassistant/components/backup_sftp/config_flow.py

Co-authored-by: Josef Zweck <josef@zweck.dev>

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+                    data=user_input,
+                )
+        else:
+            user_input = {}
```
Honestly, i agree. I've copied that from some other integration but i can see and agree that it's unnecessary. Assignment will be removed.

maretodoric committed:
Adjustments per PR comment recommendations. Added type hints, directly access `user_input` where required, remove unneeded `user_input` assignment, adjust title.

Add typing for `errors` and `placeholders`.
Access required user input values with `[]`, provide defaults in `.get` method where defaults are available.
Remove unneeded `user_input` assignment.
Adjust title to contain `username@host`.

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        errors = {}
+        placeholders = {}
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+                username=user_input.get(CONF_USERNAME),
+                password=user_input.get(CONF_PASSWORD),
+                private_key_file=user_input.get(CONF_PRIVATE_KEY_FILE),
+                backup_location=user_input.get(CONF_BACKUP_LOCATION),
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,97 @@
+class SFTPConfigEntryData:
+    """Dataclass holding all config entry data for an SFTP Backup Storage entry."""
+
+    host: str
+    port: int = 22
+    username: str
+    password: str = ""
+    private_key_file: str = ""
+    backup_location: str = ""
```
I see your point, but it makes code a little bit more readable by accessing the config properties as class attributes rather than accessing it as an item. Just my 2 cents about this, but if a deal breaker - i can re-write the code. `BackupAgentClient` already uses this class as an init argument. By changing it, i can allow it to take `ConfigEntry` as an argument and then call `async_start_reauth` from within the class itself if any authentication error occurs, to trigger the reauth process faster, rather than triggering it only during integration startup (from `async_setup_entry`).

Let me know your thoughts on this.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,349 @@
+        if (bool(config.private_key_file) is False) and (
+            bool(config.password) is False
+        ):
+            raise ConfigEntryError(
```
Makes sense, however i don't see a way to get those errors after config flow, unless someone manually changes config values after config flow.
What would be your recommendation? To create a custom Exception class, like `BackupAgentClientInitError` or to use something Home Assistant specific? Like `HomeAssistantError`?

maretodoric replied:
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,349 @@
+    async def async_delete_backup(self, backup: AgentBackup) -> None:
+        """Delete backup archive."""
+
+        assert backup is not None, (
+            "None object passed to `BackupAgentClient.async_delete_backup`."
+        )
```
Honestly, this is where my low understanding of backup system in general comes to place. I was worried that `async_delete_backup` would be called even if no backup was found from calling `backup.py:SFTPBackupAgent.async_get_backup`. And in purpose of de-cluttering `backup.py`, to reduce overall cognitive complexity, i wanted to perform assertion here.

zweckj replied:
no it will not be called. But we'll need to check during delete that the file is not gone in the meantime. In general we should rely on typing, if it's not typed with `None` we don't care too much about that case.

maretodoric replied:
Done.

maretodoric commented on homeassistant/components/backup_sftp/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "domain": "backup_sftp",
+  "name": "SFTP Backup Storage",
```
Okay, sooo, to change the name of integration to something like `sftp_storage` perhaps?

zweckj replied:
domain as well

joostlek replied:
I would indeed argue that `Backup` in the domain and name is weird

maretodoric committed:
Remvoe quotes around `SFTPConfigEntryData`

maretodoric committed:
Remove empty line after class docstring.

maretodoric committed:
Change integration title to correspond to changes made in config flow title.

maretodoric committed:
Remove unused import.

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(
+                    title=f"SFTP Backup - {user_config.username}@{user_config.host}:{user_config.port}",
```
Done

maretodoric committed:
Refactor client to use `SFTPConfigEntry` as argument instead of `SFTPConfigEntryData`.
Allow starting reauth flow directly from client.
Remove unused `get_identifier` method.
Remove raising `ConfigEntryError` from client - move the logic to `config_flow.py`
Do not assert arguments where `AgentBackup` is requested.

maretodoric committed:
Refactor to call `BackupAgentClient` with `SFTPConfigEntry` and `HomeAssistant` arguments.

maretodoric committed:
Refactor config flow to check for private key file and password presence.

maretodoric committed:
Refactor to call `BackupAgentClient` with `SFTPConfigEntry` and `HomeAssistant` arguments.

maretodoric committed:
Refactor tests to reflect changes

maretodoric committed:
Remove util, not needed anymore.

maretodoric committed:
Refactor `BackupAgentClient` class to work with metadata files rather than opening .tar archive to read `backup.json`.

maretodoric committed:
Remove unneeded log statements and try to precicely handle exceptions.

maretodoric committed:
Re-format

maretodoric committed:
Add missing argument to logger.

maretodoric committed:
Add `KeyImportError` hadnling when private key is not in correct format.

maretodoric committed:
Adapt tests to new changes.

maretodoric commented:
Requested changes were implemented. Please feel free to review and let me know your thoughts whenever you get a chance.
As for renaming from `backup_sftp` to `sftp_storage`, I'll do that if the integration is deemed ready for it.

At that point, should i just make a new PR or just rename the folder, manifest.json, all referrences to `backup_sftp` or `SFTP Backup` in docstrings, etc?

maretodoric committed:
Add "azure_storage" to resolve conflict.

maretodoric committed:
Merge branch 'dev' into dev

frenck commented:
> Requested changes were implemented. Please feel free to review and let me know your thoughts whenever you get a chance.

Don't forget to mark the PR ready for review if it is ready for that...

../Frenck

maretodoric committed:
Resolve pylint errors.

maretodoric committed:
Fix pylint errors in components directory.

zweckj commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,155 @@
+            async with BackupAgentClient(self._entry, self._hass) as client:
+                return await client.async_list_backups()
+        except Exception as e:
+            LOGGER.exception(e)
+            LOGGER.error(
+                "Listing backups failed. Please review previous exception traceback"
+            )
```
```suggestion
            LOGGER.exception("Listing backups failed.")
```
don't log both an error and a exception

zweckj commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,155 @@
+        iterator = await open_stream()
+
+        try:
+            LOGGER.debug(
+                "Establishing SFTP connection to remote host in order to upload backup"
+            )
```
can be moved out of try

zweckj commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,155 @@
+        try:
+            async with BackupAgentClient(self._entry, self._hass) as client:
+                return await client.async_list_backups()
+        except Exception as e:
```
can we be  more specific with exception captchure?

zweckj commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,247 @@
+    assert len(mock_client.mock_calls) == 1
+
+    # Verify that a new config entry is created.
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```
`is` for enum comparison (everywhere else as well)

zweckj replied:
this one is still open

maretodoric replied:
Sorry, missed this one. Done.

zweckj commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,247 @@
+    assert result["step_id"] == "user"
+
+    # Again, create a tempfile so we don't run into ConfigEntryError
+    with (
+        patch(
+            "homeassistant.components.backup_sftp.config_flow.connect",
+            return_value=fake_connect,
+        ),
+        patch(
+            "homeassistant.components.backup_sftp.config_flow.SSHClientConnectionOptions",
+            return_value=MagicMock(),
+        ),
+        NamedTemporaryFile() as tmpfile,
+    ):
```
can we rather use fixtures

maretodoric replied:
Care to provide an example? I know I'm kind of duplicating but at one place I'm raising an exception, in other place reading the number of calls to assert to fake client.
Writing test cases are kind of my bad side, i would appreciate any help here, please.

zweckj replied:
yes, but you can just set `my_fixture.return_value = ...` or `my_fixture.side_effect = ...` (and if it is otherwise the same test even do `setattr(my_fixture, "return_value" / "side_effect") = ...`. There should be plenty examples in the codebase, might wanna check lamarzocco or whitings maybe.

maretodoric replied:
Done.

zweckj commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,247 @@
+    """Test successful failure of already added config entry."""
+
+    async def add(user_input: dict):
+        config_entry.add_to_hass(hass)
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER}
+        )
+        assert result["step_id"] == "user"
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input
+        )
+        await hass.async_block_till_done()
+        return result
```
keep that inline

maretodoric replied:
Done

zweckj commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,247 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result.get("type") is FlowResultType.ABORT
+    assert result.get("reason") == reauth_reason
+    assert result.get("description_placeholders") is None
```
don't `get` on result

maretodoric replied:
Done.

zweckj commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,247 @@
+        result = await add(user_input)
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] and result["errors"]["base"] == error_base
```
for config flow tests we always want to come to a terminal state (abort/create_entry), so fix the errors and show the flow can recover and finish successfully

maretodoric replied:
Done.

zweckj commented on tests/components/backup_sftp/test_client.py:
```diff
@@ -0,0 +1,404 @@
+
+    config_entry.add_to_hass(hass)
+    cfg = SFTPConfigEntryData(**config_entry.data)
+    setattr(config_entry, "runtime_data", cfg)
```
don't patch the runtime data, instead make sure the integration loads correctly so we have something in there

maretodoric replied:
Done.

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,209 @@
+
+        return user_input
+
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Perform reauth upon connection failure."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm reauth dialog."""
+        return await self.async_step_user(user_input, "reauth_confirm")
```
Wait, i don't get it. Why would this entire piece be removed? How would i trigger reauth? I'm not currently raising `ConfigEntryAuthFailed` anywhere. By raising it, would it trigger reauth flow?

Additionally, only good place to raise it during runtime (in case credentials change before Home Assistant is restarted and before integration is reloaded) is during client initialization (`BackupAgentClient.__aenter__`), however, it was noted not to raise ConfigEntry exceptions outside of config flow.

Can you please provide some guidance on this @zweckj ?

zweckj replied:
Don't for now. The places where you raise it are ok, b this flow is not ok as is and instead of further enlarging an already huge PR let's keep re-auth for a follow up PR. But if you don't have a re-auth flow you'll need to raise `ConfigEntryError` instead of `ConfigEntryAuthFailed ` to not make it fail

maretodoric replied:
Ok, will leave it for now, then. I'm open to ideas on improving this.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,93 @@
+        raise ConfigEntryNotReady from e
+
+    # Notify backup listeners
+    _async_notify_backup_listeners_soon(hass)
```
Done

maretodoric replied:
Updated.

maretodoric committed:
Update notification of listeners to use `entry.async_on_state_change`

maretodoric committed:
Raise `BackupNotFound` when backup is not found for download.
Move logging call outside of `try` statement.
Avoid logging exceptions and error together.

maretodoric committed:
Add `setup_integration` and `setup_backup_integration` fixture to be re-used to avoid code duplication.

maretodoric committed:
Use `setup_backup_integration` from `__init__.py` to avoid code duplication.
Re-write metadata not found test to actually return 0 backups.

maretodoric committed:
Use `setup_backup_integration` from `__init__.py` to avoid code duplication.
Properly initialize the config entry rather than patching the runtime data.

maretodoric committed:
Use fixture for frequently used calls.

maretodoric committed:
Use `setup_integration` from `__init__.py` to avoid code duplication.

maretodoric committed:
Merge branch 'dev' into dev

maretodoric commented:
Hi @zweckj , I've implemented changes from your comments. Let me know if this PR should be closed and another one opened if i rename the integration to `sftp_storage` for example (since that is one changed left to be completed), or i can just rename it in this PR and of course, update the corresponding docs and icons PRs.

maretodoric committed:
Merge branch 'dev' into dev

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Setup Remote SFTP Backup Storage",
```
-> "Set up remote SFTP backup storage"

maretodoric replied:
Done, thanks for pointing out.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+          "port": "[%key:common::config_flow::data::port%]",
+          "username": "[%key:common::config_flow::data::username%]",
+          "password": "[%key:common::config_flow::data::password%]",
+          "private_key_file": "Private Key File",
```
-> "Private key file"

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+        },
+        "data_description": {
+          "host": "Hostname or IP address of SSH/SFTP Server to connect to.",
+          "port": "Port of your SSH/SFTP Server. This is usually 22.",
```
-> ""Port of your SSH/SFTP server. …"

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+          "host": "Hostname or IP address of SSH/SFTP Server to connect to.",
+          "port": "Port of your SSH/SFTP Server. This is usually 22.",
+          "username": "Username to authenticate with.",
+          "password": "Password to authenticate with. Provide this or Private Key File",
```
-> "… or private key file"

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+        }
+      },
+      "reauth_confirm": {
+        "description": "Re-authenticate with Remote SFTP Backup Storage",
```
-> "Re-authenticate with remote SFTP backup storage"

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+          "port": "[%key:common::config_flow::data::port%]",
+          "username": "[%key:common::config_flow::data::username%]",
+          "password": "[%key:common::config_flow::data::password%]",
+          "private_key_file": "Private Key File, if changed.",
```
-> "Private key file, …"

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+          "host": "Hostname or IP address of SSH/SFTP Server to connect to.",
+          "port": "Port of your SSH/SFTP Server. This is usually 22.",
+          "username": "Username to authenticate with.",
+          "password": "Password to authenticate with. Provide this or Private Key File",
```
-> "… or private key file."

maretodoric replied:
Done.

NoRi2909 commented on homeassistant/components/backup_sftp/strings.json:
```diff
@@ -0,0 +1,57 @@
+    },
+    "abort": {
+        "already_configured": "Integration already configured. Host with same address, port, username and backup location already exists.",
+        "reauth_key_changes": "Key changes were detected (Host, Port, Username or Backup location). Reauth supported only for credential changes (password or private key). For anything else, please remove integration and add new one.",
```
-> "… (host, port, username or backup location) …"

maretodoric replied:
Done.

NoRi2909 reviewed (commented):
All strings in Home Assistant should be sentence-cased.
So find a few comments below where fixes are needed.

maretodoric committed:
Convert strings to sentence-cased.

maretodoric commented:
> All strings in Home Assistant should be sentence-cased. So find a few comments below where fixes are needed.

All done, thanks for pointing both out both grammar errors and the rule that they should be sentence-cased!

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+            )
+            self.cfg.async_start_reauth(self.hass)
+            raise HomeAssistantError(e) from e
+        await self._ssh.__aenter__()
+
+        # Configure SFTP Client Connection
+        try:
+            self.sftp = await self._ssh.start_sftp_client()
+            await self.sftp.__aenter__()
```
Calls to `__aenter__` on `self._ssh` and `self.sftp` are there because both are async context managers that require initialization. This ensures they're properly set up and managed within the `BackupAgentClient` context, allowing a single `async with` to handle both in later class methods, rather than to initialize SSH connection and SFTP connection multiple times during single execution of `BackupAgentClient`.

Running `__aenter__()` on `self._ssh` doesn't automatically initialize the SFTP connection. It has to be established separately.

zweckj replied:
I looked at some of the aenters of this lib and they usually just call some other publc methods, so I'd prefer to call those directly, instead of abusing the context manager methods

maretodoric replied:
Ok, my bad. Back when i started using `asyncssh`, few years ago, those `__aenter__`s actually did some initialisation. Right now, they just return `self`. So that's cool. I'll just rename current `__aenter__` to `open` and call `open` from within `__aenter__` :) Thanks for pointing out.

maretodoric committed:
Test for condition and raise exception instead of using `assert`.
Remove reauth flow, will add it in follow up PR.
Change metadata file name suffix to `.metadata.json`.
Implement `open()` class method to `BackupAgentClient` rather than calling `__aenter__` when needed.
Add comment to explain why we call `__aenter__` on `self._ssh` and `self.sftp` inside `BackupAgentClient.__aenter__`.
Call `close` and `exit` methods of SSH and SFTP clients in ``BackupAgentClient.__aexit__`.
Fill the rest of the Integration Quality Scale.

maretodoric committed:
Re-format with ruff.

maretodoric committed:
Merge branch 'dev' into dev

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+        """
+
+        metadata_file = f"{self.cfg.runtime_data.backup_location}/.{backup.backup_id}_hass_backup_metadata.json"
+        assert await self.sftp.exists(metadata_file), (
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+        metadata: BackupMetadata = await self._load_metadata(backup)
+
+        try:
+            assert await self.sftp.exists(metadata.file_path), (
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,86 @@
+    # with SSH server or config.
+    try:
+        async with BackupAgentClient(entry, hass) as client:
+            assert isinstance(await client.list_backup_location(), list)
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,209 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm reauth dialog."""
+        return await self.async_step_user(user_input, "reauth_confirm")
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+            "metadata": backup.as_dict(),
+        }
+        async with self.sftp.open(
+            f"{self.cfg.runtime_data.backup_location}/.{backup.backup_id}_hass_backup_metadata.json",
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/quality_scale.yaml:
```diff
@@ -0,0 +1,52 @@
+    status: exempt
+    comment: No actions and no entities.
+  reauthentication-flow: done
+  test-coverage: done
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+
+    async def _initialize(self) -> None:
+        """Load file object."""
+        self._client = await BackupAgentClient(self.cfg, self.hass).__aenter__()  # pylint: disable=unnecessary-dunder-call
```
Done.

maretodoric committed:
Rename `__aenter__` to `open` and call `open` from within `__aenter__`.
Do not call `__aenter__` dunder methods of SSH Client and SFTP Client, they self initialize.

maretodoric committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,89 @@
+    private_key_file: str = ""
+    backup_location: str = ""
+
+    def __str__(self):
+        """Override string implementation of Config Entry to hide password."""
+        str_repr = super().__str__()
+        if bool(self.password):
+            return str_repr.replace(
+                f"password='{self.password}'", "password='<hidden>'"
+            )
+        return str_repr
```
we shouldn't override internal like that
```suggestion
```

maretodoric replied:
It's done, I've used `field(repr=False)` to hide the password.

zweckj commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,86 @@
+            "Failure occurred during integration setup. Reauth is needed. %s", str(e)
+        )
+        raise
+    except Exception as e:
```
is still open

maretodoric replied:
Done.

zweckj reviewed (changes requested):
there are a couple of unresolved comments from my last pass

maretodoric committed:
Hide password from printing the `SFTPConfigEntryData` class by using `field(repr=False)`.

maretodoric committed:
Redefine test that checks if password is printed in plaintext.

maretodoric commented:
> there are a couple of unresolved comments from my last pass

`__str__` override in is handled now.
As for the other comment, I'm not sure what happened but that is changed, how is the old code still displayed, i have no idea. Change is even visible in my forked repo as I'm checking now...

<img width="810" alt="image" src="https://github.com/user-attachments/assets/64c89ee9-74e2-4684-bba2-4bdac4ee96bc" />

I'm pushing new changes (not overriding `__str__`) now, maybe it will be visible.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+                "Failure while attempting to establish SSH connection. Re-auth might be required"
+            )
+            self.cfg.async_start_reauth(self.hass)
+            raise HomeAssistantError(e) from e
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        LOGGER.error(
+            "Failure occurred during integration setup. Reauth is needed. %s", str(e)
+        )
+        raise
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,160 @@
+            async with BackupAgentClient(self._entry, self._hass) as client:
+                LOGGER.debug("Uploading backup: %s", backup.backup_id)
+                await client.async_upload_backup(iterator, backup)
+        except Exception as e:
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,160 @@
+        LOGGER.debug("Received request to delete backup id: %s", backup_id)
+
+        try:
+            backup = await self.async_get_backup(backup_id)
```
Quick question. Do i raise this inside `self.async_get_backup` (so basically from backup_sftp/client.py) or to add like:
```
if backup is None:
    raise BackupNotFound(f"Backup {backup_id} not found")
```
Like it currently is?

maretodoric commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,160 @@
+                LOGGER.debug(f"Returning backup id: {backup_id}. {backup}")
+                return backup
+
+        return None
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,160 @@
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AgentBackup | None:
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+
+    def __init__(
+        self,
+        cfg: "SFTPConfigEntry",
```
This will cause [TC004](https://docs.astral.sh/ruff/rules/runtime-import-in-type-checking-block/#runtime-import-in-type-checking-block-tc004) ruff warning, because `SFTPConfigEntry` is imported as a conditional:

https://github.com/home-assistant/core/blob/30f52710788b594dc413235de07e0890aaf486e2/homeassistant/components/backup_sftp/client.py#L24-L25

If i move it outside of that conditional - it will cause recursive imports.
If i remove quotes as suggested, it will raise `NameError` during import.

maretodoric replied:
Ok, using `from __future__ import annotations`, got it :)

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+        """Load file object."""
+        self._client = await BackupAgentClient(self.cfg, self.hass).__aenter__()  # pylint: disable=unnecessary-dunder-call
+        self._fileobj = await self._client.sftp.open(self.file_path, "rb")
+        await self._fileobj.__aenter__()  # pylint: disable=unnecessary-dunder-call
```
Good point! Done.

maretodoric committed:
Refactor error handling in SFTP backup integration to improve exception specificity.

maretodoric committed:
Enhance error handling in BackupAgentClient to specifically catch possible SFTP errors and metadata reading errors.

maretodoric committed:
Ruff format.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,301 @@
+        self.sftp: SFTPClient | None = None
+        LOGGER.debug("Initialized with config: %s", self.cfg.runtime_data)
+
+    async def __aenter__(self) -> "BackupAgentClient":
```
Done.

maretodoric committed:
Refactor type hints in BackupAgentClient and AsyncFileIterator for improved clarity and consistency.

maretodoric committed:
Merge branch 'dev' into dev

zweckj commented on tests/components/backup_sftp/test_backup.py:
```diff
@@ -0,0 +1,407 @@
+    assert response["result"]["backup"] == expected_result
+
+
+@patch("homeassistant.components.backup_sftp.backup.BackupAgentClient")
```
instead of patching this everywhere, make it a fixture

maretodoric replied:
Done.

maretodoric committed:
Revert changes made accidentally.

maretodoric commented on homeassistant/components/azure_storage/quality_scale.yaml:
Looks like i accidentally made changes to wrong file. Will revert this now. As for `mypy.ini`, those changes were pulled in when syncing with base branch.

maretodoric committed:
Raise `ConfigEntryError` instead of `TypeError` during setup if an issue was encountered.

maretodoric committed:
Move `repair-issues` to `todo`.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    try:
+        async with BackupAgentClient(entry, hass) as client:
+            if not isinstance(await client.list_backup_location(), list):
+                raise TypeError("Unexpected error while setting up integration.")
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,77 @@
+        async with BackupAgentClient(entry, hass) as client:
+            if not isinstance(await client.list_backup_location(), list):
+                raise TypeError("Unexpected error while setting up integration.")
+    except (BackupAgentAuthError, RuntimeError) as e:
```
Terminal in sense that integration cannot proceed.
`RuntimeError` is raised if user renamed the directory where backups were previously stored or possibly changed ownership of directory. It will prevent integration from moving forward until those issues are resolved.

`BackupAgentAuthError` is raised if user credentials are changed. Both can be fixed with user action. However, later when reauth flow is introduced in follow-up PR, we could ask user to change the information and allow the integration to proceed.

maretodoric committed:
Will not log error and re-raise, it'll be logged by the next logger.

maretodoric committed:
Will not log error and re-raise, it'll be logged by the next logger.

maretodoric commented on homeassistant/components/backup_sftp/__init__.py:
```diff
@@ -0,0 +1,77 @@
+            if not isinstance(await client.list_backup_location(), list):
+                raise TypeError("Unexpected error while setting up integration.")
+    except (BackupAgentAuthError, RuntimeError) as e:
+        LOGGER.error(
+            "Failure occurred during integration setup. Re-adding integration might be needed. %s",
+            str(e),
+        )
```
Done.

maretodoric committed:
Will not log error and re-raise, it'll be logged by the next logger.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,329 @@
+                )
+                continue
+            except SFTPError as e:
+                LOGGER.error(
+                    "SFTP Error occurred while attempting to extract list of backups. %s",
+                    e,
+                )
```
Done.

maretodoric committed:
Rename to `BackupAgentError` to `BackupClientError`. Former was already used in `homeassistant.components.backup`.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,329 @@
+    metadata_file: str
+
+
+class BackupAgentError(Exception):
+    """Base exception for `BackupAgentClient` class."""
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,329 @@
+        try:
+            self.sftp = await self._ssh.start_sftp_client()
+        except (SFTPNoSuchFile, SFTPPermissionDenied) as e:
+            LOGGER.exception(e)
+            LOGGER.error(
+                "Failed to create SFTP client. Re-configuring integration might be required"
+            )
```
Done.

maretodoric commented on homeassistant/components/backup_sftp/backup.py:
```diff
@@ -0,0 +1,160 @@
+            async with BackupAgentClient(self._entry, self._hass) as client:
+                LOGGER.debug("Uploading backup: %s", backup.backup_id)
+                await client.async_upload_backup(iterator, backup)
+        except (BackupAgentAuthError, RuntimeError) as e:
+            raise BackupAgentError(
+                f"Failed to upload backup to remote SFTP location. Error: {e}"
+            ) from e
```
Okay, i was under the impression i should not raise `BackupAgentError` within client, but it does make sense.

maretodoric committed:
Create `backup_agent_client` fixture instead of patching `BackupAgentClient` everywhere.

maretodoric committed:
Refactor to raise `BackupAgentError` instead of `RuntimeError` and `BackupAgentAuthError`.

maretodoric committed:
Refactor to expect BackupAgentError.

maretodoric committed:
Refactor to expect BackupAgentError.

maretodoric committed:
Refactor to expect BackupAgentError.

maretodoric committed:
Use `is` to compare enums.

MatejKovacic commented:
Any info when his will be finally implemented and update pushed out?

tabrisnet commented:
> Any info when his will be finally implemented and update pushed out?

heck, how hard would it be to transform it to work as a custom component?

tabrisnet commented:
fwiw the rebase is trivial

maretodoric committed:
Merge branch 'dev' into dev

emontnemery commented:
@maretodoric the linters are failing, can you please have a look?

MatejKovacic commented:
Any update on this? This is a very desired feature...

zweckj commented:
> Any update on this? This is a very desired feature...

The status and the open points are visible for anyone, currently it's waiting on feedback from the PR creator. There is no timeline, when it's done, it's done.

maretodoric committed:
Merge branch 'dev' into dev

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,317 @@
+                **json.loads(await f.read()), metadata_file=metadata_file
+            )
+
+    async def async_delete_backup(self, backup: AgentBackup) -> None:
```
Will be done.

maretodoric commented on homeassistant/components/backup_sftp/client.py:
```diff
@@ -0,0 +1,317 @@
+    and to avoid premature closing of file and session if `BackupAgentClient` is used
+    as context manager.
+    """
+
+    def __init__(
+        self,
+        cfg: SFTPConfigEntry,
+        hass: HomeAssistant,
+        file_path: str,
+        buffer_size: int = BUF_SIZE,
+    ) -> None:
+        """Initialize `AsyncFileIterator`."""
+        self.cfg: SFTPConfigEntry = cfg
+        self.hass: HomeAssistant = hass
+        self.file_path: str = file_path
+        self.buffer_size = buffer_size
+        self._client: BackupAgentClient | None = None
+        self._fileobj: SFTPClientFile | None = None
```
Will be done. Do i do the same with `self._initialized` variable or not? And if not, why not? :) Just attempting to understand the process.

maretodoric committed:
Refactor SFTP backup methods to use backup IDs instead of backup objects and improve error handling in tests

maretodoric committed:
Provide missing type hints and apply ruff formatting.

maretodoric commented:
> @maretodoric the linters are failing, can you please have a look?

Code is updated to address PR comments and I've checked ruff and pylint - all good. Are any other linters used?

maretodoric committed:
Apply Copilot suggestions.

maretodoric committed:
Update quality scale

maretodoric committed:
Fix test case when renaming `_initialized` to `_check_initialized`

maretodoric committed:
Use unique id from config entry.

maretodoric committed:
Fix mypy linting errors.

maretodoric committed:
Fix yamllint errors.

maretodoric committed:
Properly handle `async_list_backups` failure test case.

maretodoric committed:
Fix prettier linting error.

MatejKovacic commented:
Sorry for my question (because I am not familiar with HA development and release process) - when approximately we can expect this to be released?

maretodoric committed:
Run `ruff format` after handling `async_list_backups` failure test case.

maretodoric commented:
> Sorry for my question (because I am not familiar with HA development and release process) - when approximately we can expect this to be released?

I'm currently working on fixing some linting issues. I'm in the final stage of those. The rest depends on reviewers with write access.

zweckj commented:
> when approximately we can expect this to be released?

citing from my last reply to you

> There is no timeline, when it's done, it's done.

once it’s merged it will be in the next monthly release, if it’s merged after beta has begun the month after that.

joostlek commented on tests/components/backup_sftp/test_client.py:
```diff
@@ -0,0 +1,375 @@
+
+from homeassistant.components.backup import AgentBackup, BackupAgentError
+
+# Import the classes and functions under test.
```
```suggestion
```

joostlek commented on tests/components/backup_sftp/test_client.py:
Okay so I am wondering, what is needed to test this via the backup platform? Because now the backup tests mock out the extra client abstraction, and test_client tests that extra client abstraction. So both are tested, but the danger is in that we don't test them together. I would argue that the backup tests should test the full chain, so that if we ever change a way either the abstraction is called or the ssh client is called, we confirm that it works

maretodoric replied:
My first thought was to create a temporary SSH server and go from there, however i was told that in these test cases we usually mock the library away.

maretodoric replied:
Technically speaking, if CI environment doesn't have any restrictions to spawn ports, i don't think it would cause any issues. Tweaking the CI pipeline would not be needed. `asyncssh` library that I'm using supports creation of SSH/SFTP servers. We could iterate on any available port and spawn the server in `conftest` as a global fixture. We could export some helpers to allow it to manipulate the contents of said server and then tear it down.

I realize creating such test would require devastating amount of time, especially now considering that the integration is, by my opinion 98% completed - but wouldn't mind spending an extra time on that.

Otherwise, I'm open to suggestions. Whatever it takes to contribute.

maretodoric replied:
> > I realize creating such test would require devastating amount of time
> 
> What test do you mean in this sentence?
> 

I mean to create an SSH server using asyncssh library and then test everything that way :) But i realize it's not an option.
Phew, either way it may take a good amount of time to merge test_client with test_backup. I'll try to get on it :)

joostlek commented on tests/components/backup_sftp/test_config_flow.py:
```diff
@@ -0,0 +1,197 @@
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            bool(user_input.get(CONF_PASSWORD)) is False
+            and bool(user_input.get(CONF_PRIVATE_KEY_FILE)) is False
+        ):
+            raise ConfigEntryError(
+                "Please configure password or private key file location for SFTP Backup Storage."
+            )
```
We should not use exceptions that have meaning, rather create your own for this purpose

joostlek commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+
+            # Error out if we did not find the private key file.
+            if not Path(user_input[CONF_PRIVATE_KEY_FILE]).exists():
+                raise ConfigEntryError(
+                    f"Private key file not found in provided path: `{user_input[CONF_PRIVATE_KEY_FILE]}`."
+                    " Place the key file in config or share directory "
+                    "and point to it by specifying path "
+                    "`/config/private_key` or `/share/private_key`.",
+                )
```
We shouldn't put these in the errors, rather make these rtanslatable

joostlek commented on homeassistant/components/backup_sftp/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            )
+
+        if bool(user_input.get(CONF_PRIVATE_KEY_FILE)):
+            # If full path to private key is not provided,
+            # fallback to /config as a main directory to look for pkey.
+            if not user_input[CONF_PRIVATE_KEY_FILE].startswith("/"):
+                user_input[CONF_PRIVATE_KEY_FILE] = (
+                    "/config/" + user_input[CONF_PRIVATE_KEY_FILE]
+                )
```
We should access the files securely. @farmio has something built in KNX where the user can upload their key, but I think we can be smarter with this

maretodoric committed:
Remove the redundant comment

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

maretodoric committed:
Rename integration and improve private key handling

- Create custom exceptions for handling SFTP storage errors.
- Update the integration manifest with dependencies and requirements.
- Improve tests cases to mock asyncssh library rather than client.

maretodoric committed:
Resolve pylint error.

maretodoric committed:
Remove `hass.async_block_till_done` callback.

maretodoric commented:
Pushed new changes addressing PR comments:

- Integration name changed to `sftp_storage`
- Accordingly updated docs and icons to match the new integration name.
- Replaced `private_key_file` from string to `file_upload` selector - user can now upload private key file that will be stored in storage directory of this integration.
- Added `async_remove_entry` in `__init__.py` - not requested but figured it might be needed if integration was removed and private key file is used.
- Removed tests for client.
- Changed all tests to mock `asyncssh` directly rather than mocking `BackupAgentClient`.

joostlek commented:
> Added async_remove_entry in __init__.py - not requested but figured it might be needed if integration was removed and private key file is used.

lol that is a good point

emontnemery reviewed (commented):
Looks mostly good to me, but I think the way the unique_id is calculated is not great if a reconfigure flow is added in the future where it's possible to change the hostname etc.

Also, I think a reauth/reconfigure flow allowing to change the secrets is needed, but that should probably be done in a follow-up PR.

joostlek commented:
@maretodoric Feel free to contact me (or probably Erik) via Discord. It would be very nice to get this over the finish line :)

maretodoric commented on homeassistant/components/sftp_storage/__init__.py:
```diff
@@ -0,0 +1,109 @@
+async def async_remove_entry(hass: HomeAssistant, entry: SFTPConfigEntry) -> None:
+    """Remove an SFTP Storage config entry."""
+
+    def remove_files(storage_dir: Path) -> None:
+        # For some reason, rmtree's `ignore_errors` would not ignore these exceptions
+        # So we call with contextlib.suppress
+        with contextlib.suppress(OSError, TypeError, ValueError, NotImplementedError):
+            shutil.rmtree(storage_dir)
+        LOGGER.debug("Removed storage directory for %s", entry.unique_id)
+
+    if bool(entry.data.get(CONF_PRIVATE_KEY_FILE)):
+        LOGGER.debug("Cleaning up after %s. ", entry.unique_id)
+        storage_dir = Path(hass.config.path(STORAGE_DIR))
+        await hass.async_add_executor_job(remove_files, storage_dir)
```
Ugh, true, I've forgot to add `DOMAIN` there as `paths` argument. Thank you for catching this!!

maretodoric committed:
Fix removal of entire .storage directory rather than just the domain.

maretodoric commented on homeassistant/components/sftp_storage/__init__.py:
```diff
@@ -0,0 +1,109 @@
+    private_key_file: list
+    backup_location: str
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this config entry."""
+        return slugify(
+            ".".join(
+                [
+                    self.host,
+                    str(self.port),
+                    self.username,
+                    self.backup_location,
+                ]
+            )
+        )
```
As discussed, falling back to `entry_id` for unique id.

maretodoric committed:
Refactor SFTP Storage integration: improve private key handling and cleanup logic

maretodoric committed:
Merge branch 'dev' into dev

maretodoric committed:
Update type hint for `mock_config_entry` fixture to use `Generator`

emontnemery commented on homeassistant/components/sftp_storage/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+        return self.async_show_form(
+            step_id=step_id,
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST): str,
+                    vol.Optional(CONF_PORT, default=22): int,
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Optional(CONF_PASSWORD): TextSelector(
+                        config=TextSelectorConfig(type=TextSelectorType.PASSWORD)
+                    ),
+                    vol.Optional(CONF_PRIVATE_KEY_FILE): FileSelector(
+                        FileSelectorConfig(accept="*")
+                    ),
+                    vol.Required(CONF_BACKUP_LOCATION): str,
+                }
+            ),
```
Use `add_suggested_values_to_schema` so the user doesn't have to fill in everything again.
Please make sure that works for the `FileSelector`; if it doesn't, exclude it from the suggested_values.

emontnemery commented on homeassistant/components/sftp_storage/exceptions.py:
These errors are only used to pass errors in the config flow, let's move them to `config_flow.py`

emontnemery commented on tests/components/sftp_storage/test_init.py:
```diff
@@ -0,0 +1,116 @@
+    )
+
+
+async def test_async_remove_entry(
```
Since we're being destructive when removing config entries, let's let this test start with two config entries. After the 1st entry is removed, that entries private key should not be removed. After the second config entry is removed, both key files and the containing folder should be removed.

NoRi2909 commented on homeassistant/components/sftp_storage/strings.json:
```diff
@@ -0,0 +1,37 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Set up remote SFTP Storage",
```
If we treat "SFTP Storage" as the integration name the word "remote" needs to be dropped here.
Otherwise we get a translation mismatch.

```suggestion
        "description": "Set up SFTP Storage",
```

maretodoric commented on homeassistant/components/sftp_storage/client.py:
```diff
@@ -0,0 +1,344 @@
+            dest_path = Path(hass.config.path(STORAGE_DIR, DOMAIN))
+            dest_file = dest_path / f".{ulid()}_{DEFAULT_PKEY_NAME}"
+
+            # Create parent directory
+            dest_file.parent.mkdir(exist_ok=True)
+            return str(shutil.move(file_path, dest_file))
```
Actually, looking at the `process_uploaded_file` code, it removes the whole directory where file is uploaded after exiting the context manager. So i will have to do this now but handle removal/cleanup in config flow if connection was not successful.

```
    try:
        yield file_upload_data.file_path(file_id)
    finally:
        file_upload_data.files.pop(file_id)
        shutil.rmtree(file_upload_data.file_dir(file_id))
```

emontnemery replied:
Then I'd suggest to try the connection from inside the `process_uploaded_file`, and copy to the final destination if connection succeeds

maretodoric committed:
Refactor SFTP Storage integration: improve error handling, migrate exceptions, and enhance private key management

emontnemery commented on homeassistant/components/sftp_storage/config_flow.py:
```diff
@@ -153,35 +172,68 @@ async def async_step_user(
+                vol.Required(CONF_BACKUP_LOCATION): str,
+            }
+        )
+        suggested_values = user_input.copy() if user_input else {}
+        # Exclude private key: FileSelector can't prefill uploaded files
+        suggested_values.pop(CONF_PRIVATE_KEY_FILE, None)
```
Simplify to:
```py
if user_input:
    # Exclude private key: FileSelector can't prefill uploaded files
    user_input.pop(CONF_PRIVATE_KEY_FILE)
...
    ... self.add_suggested_values_to_schema(USER_SCHEMA, user_input)
```

maretodoric commented on homeassistant/components/sftp_storage/__init__.py:
```diff
@@ -78,20 +77,50 @@ def remove_files(entry: SFTPConfigEntry) -> None:
             pkey.parent.rmdir()
+        except OSError as e:
+            if e.errno == 39:  # Directory not empty
+                leftover_files = [f.name for f in pkey.parent.iterdir()]
```
Well, that call is within `remove_files` sub-function of `async_remove_entry`.
And `remove_files` is called withing an executor

```
await hass.async_add_executor_job(remove_files, entry)
```

I like the `LOGGER.isEnabledFor` idea, will implement that, though somewhat differently due to possible exception that could be caused by iterdir, nice catch!

maretodoric commented on homeassistant/components/sftp_storage/client.py:
```diff
@@ -26,55 +23,25 @@
 
 
-def get_client_options(cfg: SFTPConfigEntryData) -> SSHClientConnectionOptions:
+def get_client_options(cfg: Mapping[str, Any]) -> SSHClientConnectionOptions:
```
I believe it's a leftover for when i was trying something different when checking private key in `save_uploaded_pkey_file`. The `SFTPConfigEntryData` was not yet initialized at that point but we did have the `user_input` from config flow so i went with it.
Should i revert back to `SFTPConfigEntryData` ?

maretodoric committed:
Refactor SFTP Storage integration: enhance logging in `async_remove_entry` and streamline data schema definition in config flow

maretodoric committed:
Resolve pylint errors

maretodoric committed:
Simplify logging and update `get_client_options` to use `SFTPConfigEntryData`

maretodoric committed:
Improve code coverage.

maretodoric committed:
Merge branch 'dev' into dev

emontnemery reviewed (commented):
Awesome! Just three nits to pick, then this is ready for merge.

maretodoric committed:
Refactor error handling in async_remove_entry and update default key name in constants

emontnemery reviewed (approved):
This is great, thanks a lot @maretodoric 👍

joostlek commented:
@maretodoric Mind sending me a message on discord so I can make sure to add you to the right channels
